Add redis database support

petergallagher opened the pull request:
A redis connection string can have a database number after the port, separated by a slash.

Currently if you try and this in the redis server strings it is not passed in to the client so you can only use database 0.

This gets the path and removes the `/` and casts it to an int as all redis dbs are integers. If the path is not a valid numeric it will set it to 0 which is the current behaviour.

acelaya commented on src/Cache/RedisFactory.php:
```diff
@@ -50,7 +50,7 @@ private function normalizeServer(string $server): array
         if (! is_array($parsedServer)) {
             throw new InvalidArgumentException(sprintf(
-                'Provided server "%s" is not a valid URL with format schema://[[username]:password@]host:port',
+                'Provided server "%s" is not a valid URL with format schema://[[username]:password@]host:port/database',
```
Since the database is not mandatory, perhaps it should be better as `[/database]`, like it's done with the credentials.

```suggestion
                'Provided server "%s" is not a valid URL with format schema://[[username]:password@]host:port[/database]',
```

acelaya commented on src/Cache/RedisFactory.php:
```diff
@@ -73,6 +73,13 @@ private function normalizeServer(string $server): array
+        unset($parsedServer['path']);
+
+        if ($database !== null) {
+            $parsedServer['database'] = (int) trim($database, '/');
```
If databases are always represented by a number, I would validate that instead of casting to `int`. Otherwise, if someone provides an invalid value by mistake, the connection will fail and the error won't be very useful.

Here, if database is not null, but it is also not a number, I would throw an error indicating it.

Extracting the whole conditional block to a private method is probably desirable, to avoid nesting conditions.

petergallagher replied:
I wasn't sure if you'd want an exception thrown in case someone already had an incorrect configuration here that was currently being ignored and would become an exception but I'll change it to throw an InvalidArgumentException instead, thanks for the feedback.

acelaya replied:
> I wasn't sure if you'd want an exception thrown in case someone already had an incorrect configuration here that was currently being ignored and would become an exception

That's actually a good point. Let's keep it then, and just add a TODO comment to change it for the next major version.

acelaya reviewed (commented):
Thanks!

I didn't even consider this situation. Have you managed to test your changes with Shlink and verify it works?

petergallagher commented:
I've changed the format of the error message and added the todo. I also applied it to my local shlink and confirmed I can write keys to different databases, and it continues to write to db0 if no database is provided.

acelaya commented:
> I've changed the format of the error message and added the todo. I also applied it to my local shlink and confirmed I can write keys to different databases, and it continues to write to db0 if no database is provided.

Thanks! 

Could you also point me to the redis documentation that explains this "indexed" database system? 

Also, is using the URI path for that the "standard" approach, or was that your idea?

petergallagher commented:
There isn't a great deal of information about the databases on the official redis website, the command you use to choose one is https://redis.io/docs/latest/commands/select/, this has a bit more information https://www.digitalocean.com/community/cheatsheets/how-to-manage-redis-databases-and-keys. It's mostly just like using a key prefix but it's more baked in to the connection so you don't have to worry about libraries you don't control using any namespace you want to limit them to which is what we were trying to do here.

Using the data from path I did because that's where parse_url puts everything after the port so I could just fetch it out without adding too much complication for a minor change. I looked at how the Symfony redis cache creates it's connection (https://github.com/symfony/symfony/blob/7.2/src/Symfony/Component/Cache/Traits/RedisTrait.php#L87) and they use parse_url to validate the connection string too but the rest of it is far more complex, but maybe it could be a good idea to use that to create a connection instead of parsing the string in this project?

acelaya commented:
> There isn't a great deal of information about the databases on the official redis website, the command you use to choose one is https://redis.io/docs/latest/commands/select/, this has a bit more information https://www.digitalocean.com/community/cheatsheets/how-to-manage-redis-databases-and-keys

Thanks! This is more than enough.

> Using the data from path I did because that's where parse_url puts everything after the port so I could just fetch it out without adding too much complication for a minor change

Yeah, I know you have to read it from the path key. I meant if you used the path in the connection URI for a reason, or just because it felt right. 

I guess it's probably a bit of both, but seems to be a sensible decision as far as I can see. 

Originally, the decision to require redis servers to be provided in the form of connection URIs is because they used to be passed verbatim to Predis, and later on I started to parse them beforehand to support other stuff. 

So one good reason to go with server-index-in-path would be if Predis would work also with that format, which I'm almost sure it would.

> I looked at how the Symfony redis cache creates it's connection (https://github.com/symfony/symfony/blob/7.2/src/Symfony/Component/Cache/Traits/RedisTrait.php#L87) and they use parse_url to validate the connection string too but the rest of it is far more complex, but maybe it could be a good idea to use that to create a connection instead of parsing the string in this project?

Symfony's logic is way too complex, and tries to outsmart you in many ways (too many for my taste), so I prefer to keep a very limited but straightforward and easy to debug custom piece of logic here.

Asking other things, Symfony sports many other connection "drivers" that's Shlink doesn't need to support.

petergallagher commented:
Oh I see what you mean about the path, this is how redis does it, so if you wanted to connect on a specific database in the cli you would do `redis-cli -u redis://shlink_redis:6379/5`.

acelaya commented on src/Cache/RedisFactory.php:
```diff
@@ -50,8 +50,9 @@ private function normalizeServer(string $server): array
-                'Provided server "%s" is not a valid URL with format schema://[[username]:password@]host:port',
+                'Provided server "%s" is not a valid URL with format %s',
                 $server,
+                'schema://[[username]:password@]host:port/[database]',
```
The brackets should enclose everything which is optional, so in this case it should be `[/database]` instead of `/[database]`.

```suggestion
                'schema://[[username]:password@]host:port[/database]',
```

acelaya commented on src/Cache/RedisFactory.php:
```diff
@@ -73,6 +74,16 @@ private function normalizeServer(string $server): array
+        unset($parsedServer['path']);
+
+        if ($database !== null) {
+            /**
+             * @todo Validate this is an integer and throw an exception if not for the next major version release
+             */
```
Probably one line is enough.

```suggestion
            // TODO For the next major version, validate this is an integer and throw an exception otherwise
```

acelaya reviewed (commented):
Just added two small requests. Looking good otherwise.

acelaya commented:
Thanks!

petergallagher commented:
No problem.
